src/deployment/index.ts

```diff
@@ -0,0 +1,137 @@
+import { defaultEntity, defaultHalHref, extractIdFromLink } from "@app/hal";
+import { WebState, schema } from "@app/schema";
+import { Deployment, LinkResponse } from "@app/types";
+
+export interface DeploymentResponse {
+  id: string;
+  status: string;
+  docker_image: string | null;
+  docker_repository_url: string | null;
+  git_repository_url: string | null;
+  git_ref: string | null;
+  git_commit_sha: string | null;
+  git_commit_url: string | null;
+  git_commit_message: string | null;
+  created_at: string;
+  updated_at: string;
+  _links: {
+    app: LinkResponse;
+    operation: LinkResponse;
+    configuration: LinkResponse;
+    image: LinkResponse;
+  };
+  _type: "deployment";
+}
+
+export const defaultDeploymentResponse = (
+  p: Partial<DeploymentResponse> = {},
+): DeploymentResponse => {
+  const now = new Date().toISOString();
+  return {
+    id: "",
+    status: "",
+    docker_image: "",
+    docker_repository_url: "",
+    git_repository_url: "",
+    git_ref: "",
+    git_commit_sha: "",
+    git_commit_url: "",
+    git_commit_message: "",
+    created_at: now,
+    updated_at: now,
+    _links: {
+      app: defaultHalHref(),
+      operation: defaultHalHref(),
+      configuration: defaultHalHref(),
+      image: defaultHalHref(),
+      ...p._links,
+    },
+    ...p,
+    _type: "deployment",
+  };
+};
+
+export const deserializeDeployment = (
+  payload: DeploymentResponse,
+): Deployment => {
+  const links = payload._links;
+
+  return {
+    id: `${payload.id}`,
+    status: payload.status,
+    dockerImage: payload.docker_image || "",
+    dockerRepositoryUrl: payload.docker_repository_url || "",
+    gitRepositoryUrl: payload.git_repository_url || "",
+    gitRef: payload.git_ref || "",
+    gitCommitSha: payload.git_commit_sha || "",
+    gitCommitUrl: payload.git_commit_url || "",
+    gitCommitMessage: payload.git_commit_message || "",
+    createdAt: payload.created_at,
+    updatedAt: payload.updated_at,
+    appId: extractIdFromLink(links.app),
+    operationId: extractIdFromLink(links.operation),
+    imageId: extractIdFromLink(links.image),
+    configurationId: extractIdFromLink(links.configuration),
+  };
+};
+
+export const selectDeploymentById = schema.deployments.selectById;
+const selectDeploymentsAsList = createSelector(
+  schema.deployments.selectTableAsList,
+  (deployments) =>
+    deployments.sort((a, b) => {
+      return new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime();
+    }),
+);
+export const selectDeploymentsByAppId = createSelector(
+  selectDeploymentsAsList,
+  (_: WebState, p: { appId: string }) => p.appId,
+  (deployments, appId) => {
+    return deployments.filter((d) => d.appId === appId);
+  },
+);
+
+export function getRegistryParts(url: string): { name: string; tag: string } {
+  const [name, tag] = url.split(":");
+  return { name, tag };
+}
+
+export function getRepoNameFromUrl(url: string): string {
+  return url.replace("https://", "");
```

Comment:
Maybe a little more robust to use a `URL` object here? Something like:

```suggestion
  const parsed = new URL(url)
  
  return [parsed.host, parsed.pathname.replace(/\.git$/, '')].join("")
```

That would give you:

```javascript
>>> const parsed = new URL("https://github.com/whatever/repo.git")
>>> [parsed.host, parsed.pathname.replace(/\.git$/, '')].join("")
'github.com/whatever/repo'

Reply:
Yeah that's basically the same thing. I was just removing the `.git` from the end to clean it up a bit, but we don't have to do that.